### ISSUE TITLE
Long messages and "Servicebus Lock Duration" (go-amqp)

### ISSUE THREAD

**Issue by Zenlil** [Azure/azure-service-bus-go#189]
(this is also registered in https://github.com/Azure/go-amqp/issues/12)

We are experiencing weird behaviour when messages take longer than half the lock-duration of the servicebus queue or topic.

Example:
Queue with a lock-duration of 30 seconds
All messages taking 20 seconds before Accept.

Expected behaviour:
All messages to be read and accepted by the servicebus

Actual behaviour:
1st message gets accepted correctly by the queue
all subsequent messaged are reported to the client as accepted, but the queue handles them like "Release()"
i.e no errors are reported from the "msg.Accept()"-call

The timing is always half of the "lock-duration", i.e setting lock-duration to max 5 minutes, messages taking >2.5 minutes behave the same as the scenario above.
and all messages taking less than half the lock-duration works as they should.

after some extended testing it seems that some message occasionally gets accepted correctly, but most messages end up as "released"

**Comment by TobiasSalzmann** [comment]
We're observing a similar issue. Even with less than slightly less than *0.5 lock duration.

**Comment by serbrech** [comment]
Hi, sorry that your issue was left unanswered.

Question: Are you using Prefetch/LinkCredit functionality?

**Comment by serbrech** [comment]
I could reliably reproduce this behavior in an integration test.

From what I can see, Although one might set prefetch to 1 (which sets linkcredit to 1 in amqp terms), we end up sending a  `Flow` frame as soon as the message is **taken** by the handler. 
If a message takes 20 seconds to be handled, the next message will sit on the receive channel, and the message lock is ticking. By the time the handler gets the message it might already have passed the lock duration and won't be handled successfully.

Here are (extended) logs from my tests exposing this behavior.

The test is enqueuing a bunch of messages before starting a Receiver.
The message handler sleeps for 20s, and then sends an Accept disposition
The message lock duration is at 30s

The logs show that more than 1 message are being transfered to the receiver while the first one is being handled :

```
14:26:34.157627 RX(Session): Transfer{Handle: 0, DeliveryID: 0, DeliveryTag: "\xad\xc8\x02\xec&\x89\x00H\x9f\v\x99\xaf\xf0\x82\x8c\xf4", MessageFormat: 0, Settled: false, More: false, ReceiverSettleMode: <nil>, State: <nil>, Resume: false, Aborted: false, Batchable: true, Payload [size]: 394}
14:26:34.157654 TX: adding handle to handlesByRemoteDeliveryID. linkCredit: 1
14:26:34.157671 RX: Transfer{Handle: 0, DeliveryID: 0, DeliveryTag: "\xad\xc8\x02\xec&\x89\x00H\x9f\v\x99\xaf\xf0\x82\x8c\xf4", MessageFormat: 0, Settled: false, More: false, ReceiverSettleMode: <nil>, State: <nil>, Resume: false, Aborted: false, Batchable: true, Payload [size]: 394}
14:26:34.157730 ec02c8ad-8926-4800-9f0b-99aff0828cf4 before push to receiver - deliveryCount : 0 - linkCredit: 1, len(messages): 0, len(inflight): 0
14:26:34.157746 ec02c8ad-8926-4800-9f0b-99aff0828cf4 after push to receiver - deliveryCount : 0 - linkCredit: 1, len(messages): 0, len(inflight): 0
14:26:34.157753 ec02c8ad-8926-4800-9f0b-99aff0828cf4 before exit - deliveryCount : 1 - linkCredit: 0, len(messages): 0
14:26:34.157760 Link Mux half: source: conn-issue189tagel8pn/Subscriptions/issue189, inflight: 0, credit: 0, deliveryCount: 1, messages: 0, maxCredit : 1, settleMode: second
14:26:34.157765 link.muxFlow(): len(l.messages):0 - linkCredit: 0 - deliveryCount: 1, inFlight: 0
14:26:34.157781 TX: Flow{NextIncomingID: <nil>, IncomingWindow: 0, NextOutgoingID: 0, OutgoingWindow: 0, Handle: 0, DeliveryCount: 1, LinkCredit: 1, Available: <nil>, Drain: false, Echo: false, Properties: map[]}
14:26:34.157805 TX(Session) - tx: Flow{NextIncomingID: 2, IncomingWindow: 100, NextOutgoingID: 0, OutgoingWindow: 100, Handle: 0, DeliveryCount: 1, LinkCredit: 1, Available: <nil>, Drain: false, Echo: false, Properties: map[]}
[2021-02-09T14:26:34+11:00] handling message ID df84d707-f88f-4c6e-85e1-ac58d2573096 - DeliveryTag : ec02c8ad-8926-4800-9f0b-99aff0828cf4
14:26:34.342910 RX(Session): Transfer{Handle: 0, DeliveryID: 1, DeliveryTag: "\x19\xc26J\xca}wC\x84*Ε\xf8\u0604\xcf", MessageFormat: 0, Settled: false, More: false, ReceiverSettleMode: <nil>, State: <nil>, Resume: false, Aborted: false, Batchable: true, Payload [size]: 394}
14:26:34.342934 TX: adding handle to handlesByRemoteDeliveryID. linkCredit: 1
14:26:34.342952 RX: Transfer{Handle: 0, DeliveryID: 1, DeliveryTag: "\x19\xc26J\xca}wC\x84*Ε\xf8\u0604\xcf", MessageFormat: 0, Settled: false, More: false, ReceiverSettleMode: <nil>, State: <nil>, Resume: false, Aborted: false, Batchable: true, Payload [size]: 394}
14:26:34.342986 4a36c219-7dca-4377-842a-ce95f8d884cf before push to receiver - deliveryCount : 1 - linkCredit: 1, len(messages): 0, len(inflight): 0
14:26:34.343001 4a36c219-7dca-4377-842a-ce95f8d884cf after push to receiver - deliveryCount : 1 - linkCredit: 1, len(messages): 0, len(inflight): 0
14:26:34.343005 4a36c219-7dca-4377-842a-ce95f8d884cf before exit - deliveryCount : 2 - linkCredit: 0, len(messages): 0
14:26:34.343011 Link Mux half: source: conn-issue189tagel8pn/Subscriptions/issue189, inflight: 0, credit: 0, deliveryCount: 2, messages: 0, maxCredit : 1, settleMode: second
14:26:34.343015 link.muxFlow(): len(l.messages):0 - linkCredit: 0 - deliveryCount: 2, inFlight: 0
14:26:34.343030 TX: Flow{NextIncomingID: <nil>, IncomingWindow: 0, NextOutgoingID: 0, OutgoingWindow: 0, Handle: 0, DeliveryCount: 2, LinkCredit: 1, Available: <nil>, Drain: false, Echo: false, Properties: map[]}
14:26:34.343048 TX(Session) - tx: Flow{NextIncomingID: 3, IncomingWindow: 100, NextOutgoingID: 0, OutgoingWindow: 100, Handle: 0, DeliveryCount: 2, LinkCredit: 1, Available: <nil>, Drain: false, Echo: false, Properties: map[]}
14:26:34.519223 RX(Session): Transfer{Handle: 0, DeliveryID: 2, DeliveryTag: "\xb3N\xeb\x00\xc5P9L\xa3\xb8&\xb3\xcfj\xff\xee", MessageFormat: 0, Settled: false, More: false, ReceiverSettleMode: <nil>, State: <nil>, Resume: false, Aborted: false, Batchable: true, Payload [size]: 394}
14:26:34.519251 TX: adding handle to handlesByRemoteDeliveryID. linkCredit: 1
```

Unwrapping the above:
The messages start to come in: 

```
RX(Session): Transfer {...}
RX: Transfer { }
```

I then added 3 log lines in the func that is pushing the message into the receiver's `messages chan Message` :

```
14:26:34.157730 ec02c8ad-8926-4800-9f0b-99aff0828cf4 before push to receiver - deliveryCount : 0 - linkCredit: 1, len(messages): 0, len(inflight): 0
14:26:34.157746 ec02c8ad-8926-4800-9f0b-99aff0828cf4 after push to receiver - deliveryCount : 0 - linkCredit: 1, len(messages): 0, len(inflight): 0
14:26:34.157753 ec02c8ad-8926-4800-9f0b-99aff0828cf4 before exit - deliveryCount : 1 - linkCredit: 0, len(messages): 0
```

I log the length of the messages channel, as it's used somewhere further. Of course, the receiver is already waiting, trying to receive from the channel, so as soon as we push the message into it, it flows through and we never see a count of 1, so long as we have handlers able to pick messages.

The next log lines are about the decision to send a `Flow` frame, and the actual frame being transmitted to the sender: 

```
14:26:34.157760 Link Mux half: source: conn-issue189tagel8pn/Subscriptions/issue189, inflight: 0, credit: 0, deliveryCount: 1, messages: 0, maxCredit : 1, settleMode: second
14:26:34.157765 link.muxFlow(): len(l.messages):0 - linkCredit: 0 - deliveryCount: 1, inFlight: 0
14:26:34.157781 TX: Flow{NextIncomingID: <nil>, IncomingWindow: 0, NextOutgoingID: 0, OutgoingWindow: 0, Handle: 0, DeliveryCount: 1, LinkCredit: 1, Available: <nil>, Drain: false, Echo: false, Properties: map[]}
```

The `Flow` frame is what updates the peer about our ability to process messages.
With `LinkCredit=1`, the peer has no credit to send us the next message. The receiver needs to send a Flow frame to give more credit to the sender. This give 1 credit to the peer, so it knows it can send us the next message. 
And that is the issue. We only just received the message, the receiver is not done. Yet, we tell the remote peer to send us the next message.

Now, I'm unsure whether this is a bug from the `go-amqp` point of view. I don't think that AMQP is supposed to wait for the disposition to be performed before sending the next Flow frame.
However, from an application developer perspective using azure service bus, my expectation with prefetch disabled (=1) is that I do not get the next message until I completed the previous one.

If the fix is in go-amqp, then the receiver should not pop the message from the channel before it is disposed, or the library should keep track of message disposal in a different structure. 
However, you can't peek at channel items, so I think it's impossible to implement this without changing go-amqp way to expose messages.

**Comment by serbrech** [comment]
@jhendrixMSFT is there anyone from the .net Servicebus SDK that could look at this issue and tell me what's the right behavior supposed to be?

**Comment by serbrech** [comment]
I think I figured out what is going on, and the go-amqp library isn't the only one at fault.

I realized that, there was a point at which amqp would see a length of 1 in the message channel, and stop sending the Flow frame, waiting for the message to be consumed downstream.

`azure-service-bus-go` implements the `listenForMessages()` by looping on amqp.Receive() and pushing the incoming message onto an unbuffered channel of messages (effectively blocking). but the message is pulled and the handlr is invoked, so the next message can be pushed in.
in effect, the intermediary channel, even if blocking unbuffered, allow for 4 messages to be transfered before the credit stop being updated.

1 as a TCP frame, blocked on pushing into the amqp message chan.
1 in the amqp messages chan
1 in the servicebus messages chan
and 1 getting processed by the handler.

Scenario: 

Each Flow frame sent to the peer triggers the peer to send a Transfer frame with the next message, which starts the message expiration lock timeout.

AMQP: 
- receives Transfer frame with message
- pushes message onto receiver message channel

SB listener loop : 
- pulls message off AMQP receiver via `Receive()` => amqp channel is now empty -> sends `Flow` :one:
- message is pushed onto SB receiver messages chan, and the handler is invoked -> SB rcv channel now empty
- listener loops, and pulls the next message off the AMQP receiver channel => amqp channel empty again, `Flow` :two: is sent.
- previous message is already in the handler, so SB message chan is empty.
- message is pushed onto SB `messages` channel and loops again
- pulls next message off AMQP message channel => amqp channel empty, `Flow` :three:
- SB now blocks on pushing the message on SB messages chan, since there is one already waiting for the handler.
- since Flow :three: was sent, a 4th message is fetched and pushed onto the message chan in amqp. this time, it's not consumed so len(l.messages) == 1, which is what prevents the Flow frame from being sent.

This is exactly the behavior I see consistently with prefetch set to 1. 
3 Flow frames are sent before seeing a length of 1 on the messages length, which equals to 4 messages buffered.

In short, this is a bug in azure-service-bus-go which is creating a buffer that equals to messages being prefetched.
However, with the current implementation of go-amqp, I think we would still see 1 message being buffered when prefetch = 1

**Comment by serbrech** [comment]
My initial hunch is that the receive pipeline should not use any channel, and should stay synchronous all the way to the user handler

**Comment by devigned** [comment]
When using peek lock (second mode), I think the Flow frame should not be sent back until the inflight message has a disposition.

If the client waited to send flow back until the message disposition was received, then that should solve the issue, right?
